File: cses/05_range_queries/1734_distinct-values-queries/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    arr = data[idx:idx + n]
    idx += n

    # Đọc truy vấn dưới dạng (b, a, k) rồi SẮP XẾP theo đầu mút phải b tăng dần.
    queries = []
    for k in range(q):
        a = int(data[idx]); b = int(data[idx + 1]); idx += 2
        queries.append((b, a, k))
    queries.sort()

    # Fenwick tree (BIT) 1-indexed trên các vị trí 1..n.
    tree = [0] * (n + 1)

    def update(i, delta):
        while i <= n:
            tree[i] += delta
            i += i & (-i)

    def prefix(i):
        s = 0
        while i > 0:
            s += tree[i]
            i -= i & (-i)
        return s

    ans = [0] * q
    last = {}          # giá trị (dạng bytes) -> vị trí xuất hiện gần nhất
    pos = 1            # vị trí tiếp theo cần thêm vào BIT (1-indexed)

    for b, a, k in queries:
        # Mở rộng tiền tố tới b, chỉ giữ dấu 1 tại lần xuất hiện cuối của mỗi giá trị.
        while pos <= b:
            v = arr[pos - 1]
            prev = last.get(v)
            if prev is not None:
                update(prev, -1)   # gỡ đánh dấu cũ
            update(pos, 1)         # đánh dấu vị trí mới
            last[v] = pos
            pos += 1
        # Số giá trị phân biệt trong [a, b] = tổng đoạn BIT.
        ans[k] = prefix(b) - prefix(a - 1)

    out = b"\n".join(str(x).encode() for x in ans)
    sys.stdout.buffer.write(out + b"\n")


main()

File: cses/05_range_queries/1734_distinct-values-queries/solution.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

static size_t to_index(int value) {
    return static_cast<size_t>(value);
}

// Fenwick tree (BIT) 1-indexed để cập nhật điểm và tính tổng tiền tố.
class FenwickTree {
public:
    explicit FenwickTree(int size) : tree_(to_index(size + 1), 0) {}

    // Cộng delta vào vị trí position.
    void add(int position, int delta) {
        for (int i = position; i < static_cast<int>(tree_.size()); i += i & -i) {
            tree_[to_index(i)] += delta;
        }
    }

    // Tổng tiền tố các vị trí 1..position.
    int prefix_sum(int position) const {
        int result = 0;
        for (int i = position; i > 0; i -= i & -i) {
            result += tree_[to_index(i)];
        }
        return result;
    }

private:
    vector<int> tree_;
};

struct Query {
    int left;
    int right;
    int index;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;

    vector<int> values(to_index(n + 1));
    for (int i = 1; i <= n; ++i) {
        cin >> values[to_index(i)];
    }

    // Đọc truy vấn rồi SẮP XẾP theo đầu mút phải tăng dần (xử lý offline).
    vector<Query> queries(to_index(q));
    for (int i = 0; i < q; ++i) {
        cin >> queries[to_index(i)].left >> queries[to_index(i)].right;
        queries[to_index(i)].index = i;
    }
    sort(queries.begin(), queries.end(), [](const Query &a, const Query &b) {
        return a.right < b.right;
    });

    FenwickTree active_positions(n);
    // last_position[x] = vị trí xuất hiện gần nhất của giá trị x.
    unordered_map<int, int> last_position;
    last_position.reserve(static_cast<size_t>(2 * n));
    last_position.max_load_factor(0.7F);

    vector<int> answers(to_index(q));
    int position = 1;
    for (const Query &query : queries) {
        // Mở rộng tiền tố tới query.right, chỉ giữ dấu 1 ở lần xuất hiện cuối.
        while (position <= query.right) {
            const auto previous = last_position.find(values[to_index(position)]);
            if (previous != last_position.end()) {
                active_positions.add(previous->second, -1);  // gỡ đánh dấu cũ
            }
            active_positions.add(position, 1);               // đánh dấu vị trí mới
            last_position[values[to_index(position)]] = position;
            ++position;
        }
        // Đếm số giá trị phân biệt trong [left, right] bằng tổng đoạn BIT.
        answers[to_index(query.index)] = active_positions.prefix_sum(query.right)
                                      - active_positions.prefix_sum(query.left - 1);
    }

    for (int answer : answers) {
        cout << answer << '\n';
    }

    return 0;
}
